### PR TITLE
Fix outdated worker address info returned by consistent hash policy

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Fix outdated worker address info returned by consistent hash policy

Summary of changes:

1. `ConsistentHashProvider` only concerns about `WorkerIdentity` when building the hash ring. Therefore, the APIs have been limited to accept and return `WorkerIdentity`s.
2. `ConsistentHashProvider.refresh` now accepts a set of worker identities instead of a list, as the order does not matter.
3. Added a test to cover the bug fix.

### Why are the changes needed?

Fix a bug where the consistent hash provider caches the `BlockWorkerInfo` of all workers, and when a worker changes its network addresses but its ID stays the same, the hash provider won't update the worker's info. A client will continue to use the outdated network address.

The fix is to make the hash provider to only consider `WorkerIdentity`s, and let the client to figure out the worker's address with the ID provided by the consistent hash provider.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

File: dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java [hunk 1]
```diff
@@ -134,6 +136,9 @@ public void testLsR() throws Exception {
     );
   }
 
+  @DoraTestTodoItem(owner = "Hua", action = DoraTestTodoItem.Action.REMOVE,
+      comment = "currently access time is not recorded and updated")
+  @Ignore
   @Test
   public void testLsWithSortByAccessTime() throws IOException, AlluxioException {
```
Comment (dbw9580) [inline dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java:143]: @huanghua78 I propose to remove this test as currently access time is not recorded and updated as part of the file metadata.

Comment (dbw9580) [reply dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java:143]: It will also be very hard to implement access time tracking since the file is always free to be accessed out of band in the UFS.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 2]
```diff
@@ -16,19 +16,27 @@
...
  * A policy where a file path is matched to worker(s) by a consistenct hashing algorithm.
  * The hash algorithm makes sure the same path maps to the same worker sequence.
  * On top of that, consistent hashing makes sure worker membership changes incur minimal
  * hash changes.
  */
 public class ConsistentHashPolicy implements WorkerLocationPolicy {
-  private static final ConsistentHashProvider HASH_PROVIDER =
+  private static final Logger LOG = LoggerFactory.getLogger(ConsistentHashPolicy.class);
+
+  private final ConsistentHashProvider mHashProvider =
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:39]: @jiacheliu3 do you remember why `HASH_PROVIDER` was a static field? The static-ness and its internal cache of worker lists are causing integration tests to fail, because workers generated in last test is carried over to next test.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 3]
```diff
@@ -57,12 +61,27 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+    List<WorkerIdentity> workers = mHashProvider.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    // todo(bowen): this is quadratic complexity. examine if it's worthwhile to replace
+    //  with an indexed map if #workers is huge
+    for (WorkerIdentity worker : workers) {
```
Comment (elega) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:76]: will there be a chance where the worker identites are not consistent with workers in blockWorkerInfo? In such case what will happen?

File: dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java (hunk 1 above)
Comment (elega) [reply dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java:143]: yeah there's a bug here and i encountered something similar

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java (hunk 3 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:76]: `mHashProvider.refresh` should update the provider's internal state to be the same as the workers in blockWorkerInfo. In cases where the provider's internal cache of the workers has not expired (TTL is 1s), it could be true that the provider is using an stale view of the cluster. In this case, this one particular call will fail. After the TTL is up, the provider should be updated to use the latest worker list.

File: dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java (hunk 1 above)
Comment (huanghua78) [reply dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java:143]: I will fix this in https://github.com/TachyonNexus/enterprise/pull/4668

Comment (dbw9580) [reply dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java:143]: thanks. For this PR, I will leave this test ignored, and you can un-ignore it when your fix lands.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 4]
```diff
@@ -57,12 +61,27 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+      for (BlockWorkerInfo info : blockWorkerInfos) {
+        if (info.getIdentity().equals(worker)) {
+          builder.add(info);
+          break;
+        }
+      }
```
Comment (JiamingMai) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:82]: We can add caching here to avoid performance degradation.

**Review by JiamingMai — APPROVED** [review]
LGTM

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 5]
```diff
@@ -57,12 +61,27 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
...
+        .collect(Collectors.toList());
+    mHashProvider.refresh(workerIdentities, mNumVirtualNodes);
+    List<WorkerIdentity> workers = mHashProvider.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    // todo(bowen): this is quadratic complexity. examine if it's worthwhile to replace
```
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:74]: or an extra `Map<WorkerIdentity, BlockWorkerInfo>` may give you `O(1)`

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java [hunk 6]
```diff
@@ -128,21 +131,24 @@ public List<BlockWorkerInfo> getMultiple(String key, int count) {
...
     // check if the worker list has expired
     if (shouldRebuildActiveNodesMapExclusively()) {
       // thread safety is valid provided that build() takes less than
       // WORKER_INFO_UPDATE_INTERVAL_NS, so that before next update the current update has been
       // finished
-      if (hasWorkerListChanged(workerInfos, mLastWorkerInfos.get())) {
-        mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
-        mLastWorkerInfos.set(workerInfos);
+      List<WorkerIdentity> lastWorkerIds = mLastWorkers.get();
+      if (!workers.equals(lastWorkerIds)) {
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:147]: A `List.equals()` is `O(n)`. However this is not more complex than the existing `hasWorkerListChanged()` so this is okay

Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:147]: however watch out `List.equals()` returns false if the two lists are of different order but same instances

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheClient.java [hunk 7]
```diff
@@ -442,10 +442,13 @@ public WorkerNetAddress getWorkerNetAddress(String path) {
...
           FileSystemContext.GetWorkerListType.LIVE) : mContext.getCachedWorkers(
           FileSystemContext.GetWorkerListType.ALL);
+      checkState(!workers.isEmpty(), "No workers available in the cluster. Lost workers %s",
+          mEnableDynamicHashRing ? "excluded" : "included");
       List<BlockWorkerInfo> preferredWorkers =
           mWorkerLocationPolicy.getPreferredWorkers(workers,
               path, mPreferredWorkerCount);
-      checkState(preferredWorkers.size() > 0);
+      checkState(!preferredWorkers.isEmpty(),
+          "Worker location policy returned no usable worker. Workers available are %s", workers);
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheClient.java:451]: same format as L445?

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java [hunk 8]
```diff
@@ -172,14 +178,15 @@ private boolean shouldRebuildActiveNodesMapExclusively() {
...
     if (mActiveNodesByConsistentHashing == null) {
       synchronized (mInitLock) {
         // only one thread should reach here
         // test again to skip re-initialization
         if (mActiveNodesByConsistentHashing == null) {
-          mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
-          mLastWorkerInfos.set(workerInfos);
+          Set<WorkerIdentity> workerIdentities = ImmutableSet.copyOf(workers);
+          mActiveNodesByConsistentHashing = build(workerIdentities, numVirtualNodes);
+          mLastWorkers.set(workers);
```
Comment (jiacheliu3) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:189]: nit: add a debug log here indicating the worker list has changed?

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java (hunk 2 above)
Comment (jiacheliu3) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java]: https://github.com/Alluxio/alluxio/pull/17753#discussion_r1258440617
I believe the intention was to make this a process-scope singleton

**Review by jiacheliu3 — APPROVED** [review]
LGTM with some comments on complexity. Complexity is a separate issue which can be addressed in another PR, but preferably in the same release. Thanks!

File: dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java [hunk 9]
```diff
@@ -180,14 +185,14 @@ public void testLsWithSortBySizeAndReverse() throws IOException, AlluxioExceptio
   }
 
   @Test
-  public void testLsWithSortByCreationTime() throws IOException, AlluxioException {
```
Comment (huanghua78) [inline dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java:183]: @dbw9580 

This test failure is not caused by this patch.
This test fails in main branch.
Why is it not discovered before?

File: dora/tests/integration/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java [hunk 10]
```diff
@@ -76,6 +76,7 @@ public static Collection<Object[]> data() {
           .setProperty(PropertyKey.WORKER_PAGE_STORE_PAGE_SIZE, Constants.KB)
           .setProperty(PropertyKey.WORKER_PAGE_STORE_SIZES, "1GB")
           .setProperty(PropertyKey.USER_NETTY_DATA_TRANSMISSION_ENABLED, true)
+          .setProperty(PropertyKey.UNDERFS_XATTR_CHANGE_ENABLED, false)
```
Comment (huanghua78) [inline dora/tests/integration/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java:79]: I propose to have a new patch for such change.
Don't mix irrelevant changes in the same patch.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (huanghua78) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:74]: I think we need a better design here.
The performance to compare such a large list is high.

If we have some version control/version number of this list, we can know if the list is changed since last time it is access.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java [hunk 11]
```diff
@@ -66,26 +68,26 @@ public class ConsistentHashProvider {
...
+   * without blocking.
+   * <p>
    * Thread safety guarantees:
    * 1. At lazy-init time, mutual exclusion is provided by `synchronized(mInitLock)`
-   *    and double-checking. At this stage it is guarded by `mInitLock`.
+   * and double-checking. At this stage it is guarded by `mInitLock`.
    * 2. After init, updating the hash ring is guarded by an optimistic lock using CAS(timestamp).
-   *    There will be no blocking but a read may see a stale ring.
-   *    At this stage it is guarded by `mLastUpdatedTimestamp`.
+   * There will be no blocking but a read may see a stale ring.
```
Comment (huanghua78) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:86]: I wonder if this is beautified automatically by some scripts or IDE?
Why such change is needed?

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java [hunk 12]
```diff
@@ -66,26 +68,26 @@ public class ConsistentHashProvider {
    * Used to compare with incoming worker list to skip the heavy build process if the worker
    * list has not changed.
    */
-  private final AtomicReference<List<BlockWorkerInfo>> mLastWorkerInfos =
-          new AtomicReference<>(ImmutableList.of());
+  private final AtomicReference<List<WorkerIdentity>> mLastWorkers =
```
Comment (huanghua78) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:71]: Question:

The `WorkerIdentity` is used to identify a worker uniquely in the hash ring.
But why is this change needed?
Can we keep the original code, and use the BlockWorkerInfo::WorkerIdentity if the `WorkerIdentity` is needed to compute and maintain the hash ring?

File: dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java (hunk 9 above)
Comment (dbw9580) [reply dora/tests/integration/src/test/java/alluxio/client/cli/fs/command/DoraLsCommandIntegrationTest.java]: reverted. these tests have been disabled by some other PRs.

File: dora/tests/integration/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java (hunk 10 above)
Comment (dbw9580) [reply dora/tests/integration/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java]: reverted

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java (hunk 11 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java]: Yeah, IDE auto formatting. Reverted.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java (hunk 12 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:71]: Only `WorkerIdentity` is needed to compute the hash ring. Anything else is out of concern of `ConsistentHashProvider`. `BlockWorkerInfo` contains volatile fields such as worker network address and cache usage information, which should not be cached by `ConsistentHashProvider`. This fix is all about removing unnecessary info from the hash ring, and let the callers from higher levels to figure out the details of a selected worker.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 13]
```diff
@@ -60,16 +64,28 @@ public List<BlockWorkerInfo> getPreferredWorkers(WorkerClusterView workerCluster
...
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    for (WorkerIdentity worker : workers) {
+      WorkerInfo workerInfo = workerClusterView.getWorkerById(worker)
+          .orElseThrow(() -> new IllegalStateException(
+              String.format("Hash provider returned a non-existent worker: %s, "
+                  + "workers currently known: %s", worker, workerIdentities)));
```
Comment (dbw9580) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:81]: This can happen since `ConsistentHashProvider` has an internal cache of the list of workers, which has a TTL of 1 s. if two concurrent calls to `mHashProvider.refresh` with different lists of workers happen within the TTL, only one will actually update the hash ring, and the other will use either a stale view of workers, or use a view from the future and cause inconsistencies with the worker info list the caller currently has.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java (hunk 6 above)
Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:147]: should change to `Set<WorkerIdentity>` instead.

Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashProvider.java:145]: changed to set instead of list.

**Comment by dbw9580** [comment]
@elega @huanghua78 @JiamingMai @jiacheliu3 this PR is now ready for review again after the `WorkerClusterView` changes are finally merged.

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 14]
```diff
@@ -60,16 +64,26 @@ public List<BlockWorkerInfo> getPreferredWorkers(WorkerClusterView workerCluster
...
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    for (WorkerIdentity worker : workers) {
+      WorkerInfo workerInfo = workerClusterView.getWorkerById(worker)
+          .orElseThrow(() -> new IllegalStateException(
+              String.format("Hash provider returned a non-existent worker: %s, "
```
Comment (elega) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:78]: i feel not comfortable just throwing an exception here. this might cause the upstream application to fail

Comment (dbw9580) [reply dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java]: updated

**Review by JiamingMai — APPROVED** [review]
LGTM

**Comment by dbw9580** [comment]
alluxio-bot, merge this please
